Fix #2968, #2996, #3758, #5332, #6469: Fixed Missing Part Resets when Failing or Cancelling Work Job

IllianiBird opened the pull request:
- Ensured `shorthandedMod` is consistently reset to 0 when part work is canceled or reinitialized.
- Addressed a missing reset for reservations.

Fix #2968
Fix #2996
Fix #3758
Fix #5332
Fix #6469

### Dev Notes
When performing a task we check whether there is sufficient Astechs to assist. If there isn't we apply a shorthanded modifier, with its magnitude increasing based on how short-handed the team is. We then store that modifier in the part, so that if the job takes multiple days we can call the original modifier. This avoids a situation where the player is shorthanded on day 1, but not on day 2, so is able to then avoid the shorthanded penalty.

The problem arises when that job either fails, or is canceled (because the attempting character doesn't have a tech skill).

At that point the modifier is locked in. In some cases we were resetting the modifier, but not all. Furthermore, in one instance we also failed to unreserve the replacement part - I predict that would then essentially make the part cease to be visible to the repair dialog. Essentially the player could have the part in their warehouse, but wouldn't be able to access it. Thankfully, the likelihood of the player running into this is fairly minimal, as it would require the person attempting the task to not have a tech skill.

This was only possible if the player GM removed the tech skill, for some reason. _Or_, more likely, they were affected by the ancient 'unskilled labor' bug that got squashed in 50.03. In that bug, if you selected multiple people and assigned them to roles en masse, we were only checking whether the first person selected was eligible for the chosen role; not everyone. Meaning that if the first selected person (say, a Dependent) was skilled in a tech skill, but someone else wasn't; that person would still be assigned the role but would then fail every check. Potentially permanently reserving parts as they went.

IllianiCBT commented:
**Addendum:** I forgot to note that this will not retroactively fix parts that already have the bug 'locked in'. Those will feature the shorthanded penalty until they are reset (generally, by the task being attempted).

psikomonkie reviewed (approved):
Nice! I notice we're doing the same sequence of clearing things for parts. I also notice there's already a method that nearly does this. 

`Part::cancelAssignment` sets the tech to null, resets overtime, resets the time spent, and sets the shorthanded mod to 0. I believe it could be modified to also cancel the reservation safely - cancel reservation is already overloaded to do nothing for parts where it shouldn't be used. 

That's part of `Part` not `IPartWork`. I'm confident that it could be added to `IPartWork`. `IPartWork` is implemented by `Part` and `PodSpace`. I believe you could either:
1. Remove `cancelAssignment()` from `Part` and add it to `IPartWork` as `default void cancelAssignment()`
2. Add an abstract `cancelAssignment` to IPartWork, set the one in `Part` to be `@Override`, and implement something for `PodSpace`. 

I think option1 would be better. 

Then replace the series of 5 methods in Campaign with just `partWork.cancelAssignment();`.

That said, I'm happy to see this fixed, and this isn't a necessity, just wanted to ask about it. Thanks!

IllianiBird commented:
Added `cancelAssignment` to `IPartWork.java`

- Added `cancelAssignment` to `IPartWork.java`
- Deprecated `cancelAssignment` in `Part.java`
- Replaced uses of deprecated class with the new default from `IPartWork.java`

psikomonkie reviewed (approved):
Nice!!
